tests/legacy-cli/e2e/tests/misc/browsers.ts

```diff
@@ -0,0 +1,63 @@
+import * as express from 'express';
+import * as path from 'path';
+import { copyProjectAsset } from '../../utils/assets';
+import { replaceInFile } from '../../utils/fs';
+import { ng } from '../../utils/process';
+
+export default async function () {
+  if (!process.env['E2E_BROWSERS']) {
+    return;
+  }
+
+  // Ensure SauceLabs configuration
+  if (!process.env['SAUCE_USERNAME'] || !process.env['SAUCE_ACCESS_KEY']) {
+    throw new Error('SauceLabs is not configured.');
+  }
+
+  await ng('build', '--prod');
+
+  // Add Protractor configuration
+  await copyProjectAsset('protractor-saucelabs.conf.js', 'e2e/protractor-saucelabs.conf.js');
+
+  // Remove browser log checks as they are only supported with the chrome webdriver
+  await replaceInFile(
+    'e2e/src/app.e2e-spec.ts',
+    'await browser.manage().logs().get(logging.Type.BROWSER)',
+    '[]',
+  );
+
+  // Workaround defect in getText WebDriver implementation for Safari/Edge
+  // Leading and trailing space is not removed
+  await replaceInFile(
+    'e2e/src/app.e2e-spec.ts',
+    '\'should display welcome message\',',
+    '\'should display welcome message\', async',
```

Comment:
I don't quite follow. How is adding `async`  related to trailing spaces?

Reply:
The test is already pseudo-async.  The “getText” call returns a promise that Protractor _magically_ awaits.  I  normalized the test so it would be easier to debug.